### PR TITLE
Better context pooling

### LINKED ISSUE #11069: [Discussion] NullReferenceException at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()
This topic came from several other:
#2806
#3219
#3219

They all a bit different, but the reason (I think) the same: concurent access to HttpContext from different threads.

Our story is pretty similar: we have project on asp.net 1.1, and we widely using HttpContext and HttpContextContextAccessor for different things (mainly for custom authorization and access to Session info), and I'm not sure if all what we're doing is really safe.

My first question is: there was a several mentions in issues above that middleware is kind of safe place, but what about ActionFilters? 
We have a pretty easy one like this:
```
public class PartialViewFilter : IActionFilter
    {
        public void OnActionExecuted(ActionExecutedContext context)
        {
            
        }

        public void OnActionExecuting(ActionExecutingContext context)
        {
            context.HttpContext.Items["_partial"] = context.HttpContext.Request.Query.Keys.Contains("partial");
        }
    }
```
and catches NullReferenceException sometimes (not very often, less than 5% requests aprox).
Exception:

> System.NullReferenceException: Object reference not set to an instance of an object.
> at Microsoft.AspNetCore.Http.Internal.DefaultHttpRequest.get_Query()
> at ****.PartialViewFilter.OnActionExecuting
> …

The most interesting part is that problem start to be occurs only when we moved from 1.0 to 1.1 (no other changes at all). 

And now my second question: is any Best Practices how to avoid this? Or at least how to detect such things?

And the last one; what about to generate more accurate error? NullReferenceException is not really nice.

### PR DESCRIPTION
**Pull request by benaadams**
Supersedes https://github.com/aspnet/AspNetCore/pull/11722 and https://github.com/aspnet/AspNetCore/pull/11721

Resolves #11069

### CODE REVIEW

**Comment by davidfowl** [comment]
- Move DefaultHttpContextFactory to Hosting, we can do better things if they are in the same assembly and we can make CreateOrInitialize internal as a result.

Loving the direction, as a result this does a couple of tings:
- Allows pooling of both the `HostContext` and `HttpContext` by extension. We can now make it a class and Kestrel will reap the benefits 😄.
- We can also keep the server decoupled from `HttpContext`, that got a bit muddled when I did the context reuse.

**Comment by benaadams** [comment]
> Move DefaultHttpContextFactory

Moved, and internaled

**Comment by benaadams** [comment]
Though the other servers aren't connection orientated, they could implement the interface
```
public interface IHostContextContainer<TContext>
{
    TContext HostContext { get; set; }
}
```
with a static bounded concurrent queue on the `IFeatureCollection` they pass and also receive the benefits of pooling

File: src/Hosting/Hosting/src/Http/DefaultHttpContextFactory.cs [hunk 1]
```diff
@@ -26,12 +27,39 @@ public DefaultHttpContextFactory(IServiceProvider serviceProvider)
...
-            var httpContext = CreateHttpContext(featureCollection);
+            var httpContext = new DefaultHttpContext(featureCollection);
+            Initialize(httpContext);
+            return httpContext;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void Initialize(DefaultHttpContext httpContext, IFeatureCollection featureCollection)
+        {
+            if (featureCollection is null)
```
Comment (davidfowl) [inline src/Hosting/Hosting/src/Http/DefaultHttpContextFactory.cs]: Remove the check from an internal method.

Comment (davidfowl) [reply src/Hosting/Hosting/src/Http/DefaultHttpContextFactory.cs]: Make it an assert.

File: src/Hosting/Hosting/src/Internal/HostingApplication.cs [hunk 2]
```diff
@@ -25,44 +27,124 @@ internal class HostingApplication : IHttpApplication<HostingApplication.Context>
         {
             _application = application;
             _diagnostics = new HostingApplicationDiagnostics(logger, diagnosticSource);
-            _httpContextFactory = httpContextFactory;
+
```
Comment (davidfowl) [inline src/Hosting/Hosting/src/Internal/HostingApplication.cs]: nit: remove this space

File: src/Http/Http.Features/src/FeatureReferences.cs [hunk 3]
```diff
@@ -108,5 +108,17 @@ public void Initalize(IFeatureCollection collection, int revision)
...
         public TFeature Fetch<TFeature>(ref TFeature cached, Func<IFeatureCollection, TFeature> factory)
             where TFeature : class => Fetch(ref cached, Collection, factory);
+
+        private static int ContextDisposed()
+        {
+            ThrowContextDisposed();
+            return 0;
+        }
+
+        private static void ThrowContextDisposed()
```
Comment (davidfowl) [inline src/Http/Http.Features/src/FeatureReferences.cs]: This part of the stack unfortunately can't assume the feature collection is HttpContext based.

Comment (benaadams) [reply src/Http/Http.Features/src/FeatureReferences.cs]: Changed

**Review by davidfowl — APPROVED** [review]
LGTM

**Pull request merged**